File: bot/commands/createBossQuestionnaireCommand.py

from discord.message import Message
from command_spec import MessageCommand, PhraseRepository
import re


class CreateBossQuestionnaireCommand(MessageCommand):
    def __init__(self, phraseRepository: PhraseRepository) -> None:
        super().__init__(phraseRepository)

    def isMatchTo(self, message: str):
        return re.search(self.phraseRepository.get('create_boss_questionnaire'), message)

    async def execute(self, message: Message) -> None:
        print('start create boss querstionaire command')
        sentMessage: Message = await message.channel.send(self.phraseRepository.get('boss_questionnaire_message'))
        reactions = [
            self.phraseRepository.get('1_boss_stamp'),
            self.phraseRepository.get('2_boss_stamp'),
            self.phraseRepository.get('3_boss_stamp'),
            self.phraseRepository.get('4_boss_stamp'),
            self.phraseRepository.get('5_boss_stamp'),
        ]
        for reaction in reactions:
            await sentMessage.add_reaction(reaction)

File: bot-js/src/commands/CreateBossQuestionnaireCommand.ts
import { Client, Message } from 'discord.js';
import { MessageCommand } from './MessageCommand';
import { PhraseRepository } from '../support/PhraseRepository';
import { isTextChannel } from '../support/DiscordHelper';
import { mentionedToMe } from '../Sylph';
import { PhraseKey } from '../support/PhraseKey';
import { ApiClient } from '../backend/ApiClient';
import { pipe } from 'fp-ts/lib/function';
import * as TaskOption from 'fp-ts/lib/TaskOption';
import { roleMension } from '../support/Memtion';

export class CreateBossQuestionnaireCommand implements MessageCommand {
    constructor(
        private phraseRepository: PhraseRepository,
        private discordClient: Client,
        private apiClient: ApiClient,
    ) {
        this.commandPattern = new RegExp(this.phraseRepository.get(PhraseKey.createBossQuestionnaire()));
    }

    private readonly commandPattern: RegExp;

    async isMatchTo(message: Message): Promise<boolean> {
        const cooperateChannel = await this.apiClient.getCooperateChannel(message.channelId);

        if (cooperateChannel) {
            return Promise.resolve(
                this.commandPattern.test(message.cleanContent)
                && mentionedToMe(message, this.discordClient)
            );
        } else {
            return false;
        }
    }

    async execute(message: Message): Promise<void> {
        console.log("start create boss questionnaire command");

        if (!isTextChannel(message.channel)) return;

        await pipe(
            TaskOption.fromTask(() => this.apiClient.getCooperateChannel(message.channelId)),
            TaskOption.chainNullableK(cooperateChannel => cooperateChannel),
            TaskOption.chainTaskK(cooperateChannel => () => this.apiClient.getUncompleteMemberRole(cooperateChannel.clanId)),
            TaskOption.chainNullableK(role => role),
            TaskOption.fold(
                () => TaskOption.some(this.phraseRepository.get(PhraseKey.bossQuestionnaireMessage())),
                role => TaskOption.some(`${roleMension(role.role.discordRoleId)} ${this.phraseRepository.get(PhraseKey.bossQuestionnaireMessage())}`)
            ),
            TaskOption.chainTaskK(messageText => () => message.channel.send(messageText)),
            TaskOption.chainTaskK(sentMessage => async () => {
                await sentMessage.react(this.phraseRepository.get(PhraseKey.bossStamp(1)));
                await sentMessage.react(this.phraseRepository.get(PhraseKey.bossStamp(2)));
                await sentMessage.react(this.phraseRepository.get(PhraseKey.bossStamp(3)));
                await sentMessage.react(this.phraseRepository.get(PhraseKey.bossStamp(4)));
                await sentMessage.react(this.phraseRepository.get(PhraseKey.bossStamp(5)));
            })
        )();
    }
}